spec/02-integration/03-db/12-dao_hooks_spec.lua

```diff
@@ -0,0 +1,280 @@
+local helpers = require "spec.helpers"
+local hooks = require "kong.hooks"
+
+
+for _, strategy in helpers.each_strategy() do
+  describe("kong.db hooks [#" .. strategy .. "]", function()
+    local db, bp, s1, r1
+
+    lazy_setup(function()
+      bp, db = helpers.get_db_utils(strategy, {
+        "routes",
+        "services",
+      })
+
+      s1 = assert(bp.services:insert {
+        name = "s1",
+        url = "http://example.test",
+      })
+
+      r1 = assert(bp.routes:insert {
+        protocols = { "http" },
+        hosts = { "host1" },
+        service = s1,
+        name = "r1",
+      })
+    end)
+
+    describe("page_for", function()
+      local pre_hook = spy.new(function() end)
+      local post_hook = spy.new(function() end)
+
+      lazy_setup(function()
+        hooks.register_hook("dao:page_for:pre", function()
+          pre_hook()
+          return true
+        end)
+        hooks.register_hook("dao:page_for:post", function()
+          post_hook()
+          return true
+        end)
+      end)
+
+      lazy_teardown(function()
```

Comment:
Ensure this happens even if errors are raised
```suggestion
      finally(function()
```

Reply:
In what contexts is `finally` available?